File: src/content/docs/topics/cs/leetcode-150/graphs/207-course-schedule.py
def can_finish(num_courses, prerequisites):
    pass  # TODO: implement

def _run_tests():
    assert can_finish(2, [[1, 0]]) == True
    assert can_finish(2, [[1, 0], [0, 1]]) == False
    assert can_finish(5, []) == True
    assert can_finish(1, []) == True
    assert can_finish(3, [[1, 0], [2, 1], [0, 2]]) == False
    assert can_finish(4, [[1, 0], [2, 0], [3, 1], [3, 2]]) == True
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    can_finish(200, [[i + 1, i] for i in range(199)])
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf can-finish 200 courses DAG chain: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/graphs/207-course-schedule.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:ed8fd2c6554df64a34728495c5a678faa1d9c8dfd7659ef5258e965ff11d7621
    expectEqual(Solution().canFinish(3, [[1, 0], [2, 1]]), true, "simple-chain")
    expectEqual(Solution().canFinish(2, [[1, 0], [0, 1]]), false, "two-course-cycle")
    expectEqual(Solution().canFinish(1, []), true, "one-course")
    // EXCLUDED_VECTOR course-out-of-range: [2,[[2,0]]] | Every course identifier must be smaller than numCourses.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
